### PR TITLE
rgw: add unit tests for the external ceph object store reconciler

### LINKED ISSUE #9025: rgw: add unit tests for the extrernal ceph object store reconciler
`pkg/operator/ceph/object/controller_test.go` does not have any test for  an external CephObjectStore CR.

### PR DESCRIPTION
**Pull request by rexagod**
**Description of your changes:** Added two test cases to check creation and deletion of external object stores.

**Which issue is resolved by this Pull Request:**
Resolves #9025 

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/object/controller_test.go [hunk 1]
```diff
@@ -745,6 +791,68 @@ func TestCephObjectStoreControllerMultisite(t *testing.T) {
 		assert.True(t, dependentsChecked)
 		assert.True(t, calledCommitConfigChanges)
 	})
+
+	t.Run("create an external object store", func(t *testing.T) {
+		objects := []runtime.Object{
+			objectZone,
+			objectZoneGroup,
+			objectRealm,
+			cephCluster,
```
Comment (thotz) [inline pkg/operator/ceph/object/controller_test.go:800]: Do objectZone, objectZoneGroup, objectRealm applicable to external rgw cluster?

Comment (rexagod) [reply pkg/operator/ceph/object/controller_test.go:798]: Removed.

File: pkg/operator/ceph/object/controller_test.go [hunk 2]
```diff
@@ -615,12 +616,30 @@ func TestCephObjectStoreControllerMultisite(t *testing.T) {
...
+			Name:      store,
+			Namespace: namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind: "CephObjectStore",
+		},
+		Spec: cephv1.ObjectStoreSpec{},
+	}
+
+	externalObjectStore.Spec.Zone.Name = zoneName
```
Comment (thotz) [inline pkg/operator/ceph/object/controller_test.go:630]: `Spec.Zone` is not needed for external RGW server

**Review by thotz — CHANGES_REQUESTED** [review]
In my opinon, you should not make this changes in `TestCephObjectStoreControllerMultisite()`, u need to add new function `TestCephObjectExternalStoreControllerMultisite()` with your changes. You can refer `TestCephObjectStoreController` to implement that, you unit tests need to cover sanity for external object store, like what if `External RGW Endpoint` is null and `rgwAdminOps` Secret is missing along with normal creation

File: pkg/operator/ceph/object/controller_test.go (hunk 2 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/object/controller_test.go:630]: @thotz What should be the behavior of an external object store if a user does set this though? Can we validate that Rook does the right thing with it?

Comment (thotz) [reply pkg/operator/ceph/object/controller_test.go:630]: @BlaineEXE : From code reading, currently `ObjectStore.Spec.Zone` is used by `reconcileMultisiteCRs` which is not called if the cluster is external. And there is no validation check present in `external` store path as well.

**Review by BlaineEXE — APPROVED** [review]
I think this looks good. Thanks @rexagod

File: pkg/operator/ceph/object/controller_test.go [hunk 3]
```diff
@@ -754,6 +755,259 @@ func TestCephObjectStoreControllerMultisite(t *testing.T) {
 	})
 }
 
+func TestCephObjectExternalStoreControllerMultisite(t *testing.T) {
```
Comment (thotz) [inline pkg/operator/ceph/object/controller_test.go:758]: Again As I mentioned before Multisite(Zone/Realm/Zonegroup) configuration are not supported for `ExternalStore`.
Hence I request to check `TestCephObjectStoreController()` than checking `TestCephObjectStoreControllerMultisite()`, last by my mistake I mentioned new function name as ` TestCephObjectExternalStoreControllerMultisite()` it should just be  `TestCephObjectExternalStoreController`

File: pkg/operator/ceph/object/controller_test.go [hunk 4]
```diff
@@ -754,6 +755,259 @@ func TestCephObjectStoreControllerMultisite(t *testing.T) {
...
+		objects := []runtime.Object{
+			cephCluster,
+			externalObjectStore,
+		}
+		r := getReconciler(objects)
+		res, err := r.Reconcile(ctx, req)
+		assert.Error(t, err)
+		assert.False(t, res.Requeue)
+		assert.False(t, calledCommitConfigChanges)
+		err = r.client.Get(ctx, req.NamespacedName, externalObjectStore)
```
Comment (thotz) [inline pkg/operator/ceph/object/controller_test.go:986]: Sorry, I am a bit confused here, do `get` need to return the error here? Since  `externalObjectStore` reconcile failed. Same for 1003.

File: pkg/operator/ceph/object/controller_test.go [hunk 5]
```diff
@@ -754,6 +755,259 @@ func TestCephObjectStoreControllerMultisite(t *testing.T) {
...
+			Namespace: namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind: "CephObjectStore",
+		},
+		Spec: cephv1.ObjectStoreSpec{},
+	}
+
+	externalObjectStore.Spec.Gateway.Port = 81
+	externalObjectStore.Spec.Gateway.ExternalRgwEndpoints = []v1.EndpointAddress{{IP: ""}}
```
Comment (thotz) [inline pkg/operator/ceph/object/controller_test.go:818]: Is this bit weird @leseb, empty IP address works but `ExternalRgwEndpoints` is nil. Should we handle empty IP address as well(may be different PR)

**Review by thotz — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/object/controller_test.go (hunk 4 above)
Comment (rexagod) [reply pkg/operator/ceph/object/controller_test.go:986]: I think `get` returns `nil` as the error value here since `externalObjectStore` is still present in the runtime environment?

https://github.com/rook/rook/pull/9710/files#diff-10bcb1ddae3b1fddd221dda0765d8dc3987af28dae3a1944dbeaadc31da531fdR979

File: pkg/operator/ceph/object/controller_test.go (hunk 3 above)
Comment (rexagod) [reply pkg/operator/ceph/object/controller_test.go:758]: Ah, makes sense. I'll rename the function.

File: pkg/operator/ceph/object/controller_test.go (hunk 4 above)
Comment (thotz) [reply pkg/operator/ceph/object/controller_test.go:986]: okay

File: pkg/operator/ceph/object/controller_test.go [hunk 6]
```diff
@@ -754,6 +755,259 @@ func TestCephObjectStoreControllerMultisite(t *testing.T) {
...
+			}
+			if args[0] == "osd" && args[1] == "pool" && args[2] == "get" {
+				return "", errors.New("test pool does not exit yet")
+			}
+			if args[0] == "versions" {
+				return dummyVersionsRaw, nil
+			}
+			return "", nil
+		},
+		MockExecuteCommandWithTimeout: func(timeout time.Duration, command string, args ...string) (string, error) {
```
Comment (thotz) [inline pkg/operator/ceph/object/controller_test.go:847]: As I mentioned before multsite configs are not need line Number 847 to 863

File: pkg/operator/ceph/object/controller_test.go [hunk 7]
```diff
@@ -754,6 +755,259 @@ func TestCephObjectStoreControllerMultisite(t *testing.T) {
 	})
 }
 
+func TestCephObjectExternalStoreController(t *testing.T) {
+	ctx := context.TODO()
+	capnslog.SetGlobalLogLevel(capnslog.DEBUG)
+	os.Setenv("ROOK_LOG_LEVEL", "DEBUG")
+
+	zoneGroupName := "zonegroup-a"
```
Comment (thotz) [inline pkg/operator/ceph/object/controller_test.go:763]: not needed

File: pkg/operator/ceph/object/controller_test.go [hunk 8]
```diff
@@ -754,6 +755,259 @@ func TestCephObjectStoreControllerMultisite(t *testing.T) {
...
+	realmName := "realm-a"
```
Comment (thotz) [inline pkg/operator/ceph/object/controller_test.go:764]: not needed

File: pkg/operator/ceph/object/controller_test.go [hunk 9]
```diff
@@ -754,6 +755,259 @@ func TestCephObjectStoreControllerMultisite(t *testing.T) {
...
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "rook-ceph-mon",
+			Namespace: namespace,
+		},
+		Data: secrets,
+		Type: k8sutil.RookType,
+	}
+
+	objectZoneGroup := &cephv1.CephObjectZoneGroup{
```
Comment (thotz) [inline pkg/operator/ceph/object/controller_test.go:793]: not needed line 793 to 801

File: pkg/operator/ceph/object/controller_test.go [hunk 10]
```diff
@@ -754,6 +755,259 @@ func TestCephObjectStoreControllerMultisite(t *testing.T) {
...
+				return userCreateJSON, nil
+			}
+			return "", nil
+		},
+	}
+
+	commitConfigChangesOrig := commitConfigChanges
+	defer func() { commitConfigChanges = commitConfigChangesOrig }()
+
+	// make sure joining multisite calls to commit config changes
```
Comment (thotz) [inline pkg/operator/ceph/object/controller_test.go:870]: multisite calls are not performed here, not needed 870-874

**Review by thotz — CHANGES_REQUESTED** [review]

**Comment by rexagod** [comment]
@thotz I've committed the latest review changes in a different `fixup!` commit to make it easier to see the changes. PLMK if they look good to you and I'll squash.

**Review by thotz — APPROVED** [review]

**Review by thotz — APPROVED** [review]
LGTM

File: pkg/operator/ceph/object/controller_test.go (hunk 4 above)
Comment (leseb) [reply pkg/operator/ceph/object/controller_test.go:937]: What do we need this `get` for?

File: pkg/operator/ceph/object/controller_test.go [hunk 11]
```diff
@@ -754,6 +755,209 @@ func TestCephObjectStoreControllerMultisite(t *testing.T) {
...
+		objects := []runtime.Object{
+			cephCluster,
+			externalObjectStore,
+			rgwAdminOpsUserSecret,
+		}
+		r := getReconciler(objects)
+		res, err := r.Reconcile(ctx, req)
+		assert.NoError(t, err)
+		assert.False(t, res.Requeue)
+		err = r.client.Get(ctx, req.NamespacedName, externalObjectStore)
```
Comment (leseb) [inline pkg/operator/ceph/object/controller_test.go:890]: What do we need this `get` for?

File: pkg/operator/ceph/object/controller_test.go (hunk 4 above)
Comment (rexagod) [reply pkg/operator/ceph/object/controller_test.go]: Redundant, removed.

File: pkg/operator/ceph/object/controller_test.go (hunk 11 above)
Comment (rexagod) [reply pkg/operator/ceph/object/controller_test.go]: Redundant, removed (and other such instances).

File: pkg/operator/ceph/object/controller_test.go (hunk 5 above)
Comment (rexagod) [reply pkg/operator/ceph/object/controller_test.go:802]: Will push a PR targeting #9909.

**Review by leseb — APPROVED** [review]

**Comment by rexagod** [comment]
Squashed.
